Upgrade to lucene-9.1.0-snapshot-1336263051c

mayya-sharipova opened the pull request:
Lucene issues that resulted in elasticsearch changes:

LUCENE-9820 Separate logic for reading the BKD index from logic to intersecting it.
LUCENE-10377: Replace 'sortPos' with 'enableSkipping' in SortField.getComparator()
LUCENE-10301: make the test-framework a proper module by moving all test
classes to org.apache.lucene.tests
LUCENE-10300: rewrite how resources are read in ukrainian morfologik analyzer:
LUCENE-10054 Make HnswGraph hierarchical

mayya-sharipova committed:
Upgrade to lucene-9.1.0-snapshot-e4ad3c794fc

Lucene issues that resulted in elasticsearch changes:

LUCENE-9820 Separate logic for reading the BKD index from logic to intersecting it.
LUCENE-10377: Replace 'sortPos' with 'enableSkipping' in SortField.getComparator()
LUCENE-10301: make the test-framework a proper module by moving all test
classes to org.apache.lucene.tests
LUCENE-10300: rewrite how resources are read in ukrainian morfologik analyzer:
LUCENE-10054 Make HnswGraph hierarchical

mayya-sharipova commented:
@iverase  Looking for feedback on new `PointValues` interface in `ExitableDirectoryReader.java` and `RewriteCachingDirectoryReader.RewriteCachingLeafReader`

@romseygeek Looking for feedback on `UkrainianAnalyzerProvider` and `XUkrainianMorfologikAnalyzer.java`.

Thanks a lot in advance.

mayya-sharipova committed:
Fix AnalysisFactoryTestCase

iverase reviewed (commented):
Changes in `ExitableDirectoryReader.java` and `RewriteCachingDirectoryReader.RewriteCachingLeafReader` LGTM

romseygeek commented on plugins/analysis-ukrainian/src/main/java/org/elasticsearch/plugin/analysis/ukrainian/UkrainianAnalyzerProvider.java:
```diff
@@ -23,7 +22,7 @@
     public UkrainianAnalyzerProvider(IndexSettings indexSettings, Environment env, String name, Settings settings) {
         super(indexSettings, name, settings);
         analyzer = new XUkrainianMorfologikAnalyzer(
```
I think this is something we should fix upstream, it is a pretty clear bug in the API that you can't access the default stopword set

romseygeek replied:
I opened https://issues.apache.org/jira/browse/LUCENE-10413

mayya-sharipova replied:
@romseygeek This makes sense to me. I will update this PR with a new lucene snapshot, once your changes are merged in 9.x branch.

mayya-sharipova replied:
@romseygeek  Thanks Alan, I've upgraded to the new lucene snapshot and removed `XUkrainianMorfologikAnalyzer` in favour of Lucene's one  in ad97497abfd

mayya-sharipova committed:
Upgrade to lucene-9.1.0-snapshot-ad97497abfd

- Delete XUkrainianMorfologikAnalyzer now that UkrainianMorfologikAnalyzer
provides default stop words tests

mayya-sharipova committed:
Remove skipping of test testSimpleIndexSortEarlyTerminate
after upgrade to Lucene 9.1

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into upgrade-lucene-e4ad3c794fc

mayya-sharipova commented:
@romseygeek With these new Lucene changes I pulled  `HighlighterWithAnalyzersTests`:: `testPhrasePrefix` fails consistently.  Do you know what Lucene change could cause change in our highlighters' behaviour?

romseygeek commented:
The failure is specifically with highlighting phrase prefix queries, but I can't see anything in the recent changes that would obviously explain this.  There's a change to how interval prefixes are built, but MatchPhrasePrefix uses spans instead of intervals so I don't think that would be the issue.

mayya-sharipova committed:
TimeSeriesDataStreamsIT and ShrinkActionIT update tests

mayya-sharipova committed:
Upgrade to lucene-9.1.0-snapshot-1336263051c

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into upgrade-lucene-e4ad3c794fc

romseygeek reviewed (approved):
LGTM, thanks @mayya-sharipova

jtibshirani commented:
@mayya-sharipova @romseygeek I looked into the `HighlighterWithAnalyzersTests` failure as part of trying to upgrade to a new 9.1 snapshot. Here's my understanding of the problem: `UnifiedHighlighter` extracts terms from the query to decide what it needs to store in its `MemoryIndex`. To do this, it first [rewrites the query using an empty searcher](https://github.com/apache/lucene/blob/branch_9_0/lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/UnifiedHighlighter.java#L149), before extracting terms. Because of the new rewrite rules in https://github.com/apache/lucene/pull/664, the rewrite returns `MatchNoDocsQuery` and no terms can be extracted. This means the highlighter doesn't think that any documents could match.

We seem to have run into this before, and added a hack in the relevant query, but it's no longer working because of the stronger rewrite rules: https://github.com/elastic/elasticsearch/blob/e1202dd00e3e10d1e4bf93638989fd3055ee4627/server/src/main/java/org/elasticsearch/common/lucene/search/MultiPhrasePrefixQuery.java#L162-L169

I am unsure what to do. We could tweak the hack in `MultiPhrasePrefixQuery`, but I wonder if the problem could surface with other queries too, that also rewrite to `MatchNoDocsQuery` with an empty searcher. I don't fully understand `UnifiedHighlighter` and my analysis could be a bit off.

Also tagging @jpountz as he introduced the new rewrite rules.

romseygeek commented:
I think we can detect if the IndexReader passed to `rewrite()` is empty, and return a special ES-specific MatchNoDocsQuery that wouldn't be detected by the boolean query rewrite rules.  Empty reader detection means that this code path would only be triggered by the UnifiedHighlighter.

jtibshirani commented:
@romseygeek thanks again for taking a look! That makes sense to me and seems to fit well with the existing logic.

I was wondering if you think we can just fix this in Elasticsearch or if we should raise it as a Lucene issue. I tried to reproduce a similar failure fully within Lucene but did not see a way to do it. I couldn't find any Lucene queries that rewrote to `MatchNoDocsQuery` when there is an empty reader (for example because it consults the reader to see what terms exist). In fact in one place we seem to explicitly avoid using `MatchNoDocsQuery` and instead used a `TermQuery` that we know will not match. Maybe we could file an issue just noting that this `UnifiedHighlighter` behavior is fragile / trappy, but not consider it a blocker?

jtibshirani commented:
>  I'm pretty sure that we can reproduce the behaviour of almost everything with some query rewriting tricks.

Great, sounds like we're on the same page. To clarify, I meant I couldn't identify any existing Lucene queries that would trigger the issue (like our custom Elasticsearch queries).
